### PR TITLE
8342808: Javadoc should add whitespace between type parameters

### PR DESCRIPTION
**Pull request by nizarbenalla**
EDIT: Here are the [generated docs](https://cr.openjdk.org/~nbenalla/GeneratedDocs/8342808/docs/api/index.html) after the latest changes.

Please review this patch to render javadocs as `<K, V, T>` rather than `<K,V,T>` to match naming conventions and make the rendered output slighly nicer to read.

Passes langtool tests.

The [generated docs](https://cr.openjdk.org/~nbenalla/GeneratedDocs/K-V-space/docs/api/index.html) only include `java.base`.

Note for reviewers:

In `TestInheritence`, B is a user defined class and `TypeMirror::getKind` returns `DECLARED `. Which why we see this output. 

html
```
Class D<R,S>
java.lang.Object
[pkg.A](https://htmledit.squarefree.com/A.html)<S, [B](https://htmledit.squarefree.com/B.html)>
[pkg.B](https://htmledit.squarefree.com/B.html)<S, [B](https://htmledit.squarefree.com/B.html)>
pkg.D<R,S>

```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342808](https://bugs.openjdk.org/browse/JDK-8342808): Javadoc should add whitespace between type parameters (**Bug** - P4)

### Reviewers
 * [Hannes Wallnöfer](https://openjdk.org/census#hannesw) (@hns - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21759/head:pull/21759` \
`$ git checkout pull/21759`

Update a local copy of the PR: \
`$ git checkout pull/21759` \
`$ git pull https://git.openjdk.org/jdk.git pull/21759/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21759`

View PR using the GUI difftool: \
`$ git pr show -t 21759`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21759.diff">https://git.openjdk.org/jdk/pull/21759.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21759#issuecomment-2443938506)
</details>

### CODE REVIEW

**Comment by hns** [comment]
I'm not sure about this change. For the common case of single uppercase letter parameter names, I actually find the notation without space character more readable, and that also seems to be the convention in most OpenJDK code. 

If we are going to introduce a space character after the comma for type parameters, we also should get rid of the `<wbr>` (line break opportunity) element, as that is not needed anymore with the addes space.

**Comment by nizarbenalla** [comment]
I don't have a big preference but the default source formatting recommendations seem to use spaces?
https://github.com/openjdk/jdk/blob/3ccd2f757d069c16147b331a90b1590c8ad4ae24/src/java.base/share/classes/java/util/function/Function.java#L41

**Comment by pavelrappo** [comment]
> I'm not sure about this change.

IIRC, this feature was suggested by @minborg elsewhere. It would be nice if he advocated for it. Per, do you have anything to say here? Maybe other core-libs people have opinions on it?

> For the common case of single uppercase letter parameter names, I actually find the notation without space character more readable, and that also seems to be the convention in most OpenJDK code.

We use a mix of styles in code. However, in javadoc output, we should be uniform similarly to how we are uniform with method and constructor parameters, which we always separate with whitespace.

**Comment by pavelrappo** [comment]
Not all reviewers might realise it, but javadoc currently outputs generics both ways, with whitespace and without. For example, have a look at [Map.of]. The first occurrence of generic parameters (i.e. declaration), immediately after the `static` keyword, has whitespace. The second occurrence (i.e. use) does not. I think we need to make up our mind on what we should do.

[Map.of]: https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/util/Map.html#of()

**Comment by jonathan-gibbons** [comment]
It might be interesting to audit the JDK codebase for prevailing patterns in source code.  While regexes are not great for parsing, they may be good enough for a reasonably good analysis.
For example, as a start, look for
* `<[^> ]+>` -- instances of type parameters or arguments with no space (yes, it doesn't, do nested types, but it's good enough)
* `<[^>]+ [^>]+>` --  instances of type parameters or arguments with at least one space
* `((class|interface) [A-Za-z0-9]+)|public|static) *<[^> ]+>` -- instances of type parameters with no space
* `((class|interface) [A-Za-z0-9]+)|public|static) <[^>]+ [^>]+>` -- instances of type parameters at least one space

The regexes are just a suggestion and may need refinement.  The intent is to catch character sequences between `<` and `>` , and to further filter those that look like generic type or method declarations

**Comment by minborg** [comment]
Some of the more prominent users of more-than-one-type-parameter declaration are `Map` and `Function` where the upper-case parameters have a space separator.  I think all `java.util.function` constructs with more than one type parameter have spaces after the coma(s).  Ultimately, it is a matter of taste but at least we should be consistent.

**Comment by jonathan-gibbons** [comment]
I note there is an entity for a "thin space", U+2009. Maybe that could be a compromise here.

https://www.compart.com/en/unicode/U+2009

**Comment by hns** [comment]
> Not all reviewers might realise it, but javadoc currently outputs generics both ways, with whitespace and without. For example, have a look at [Map.of](https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/util/Map.html#of()). The first occurrence of generic parameters (i.e. declaration), immediately after the `static` keyword, has whitespace. The second occurrence (i.e. use) does not. I think we need to make up our mind on what we should do.

I just looked into this, and what actually happens is that we have a special `HtmlLinkInfo.addLineBreaksInTypeParameters` flag to insert a newline character after the comma just for this particular case, but the CSS for this particular span of the method signature uses `white-space: normal;`, so the newline character is displayed as space. 

The code to add the newline in this particular position is rather ancient (previously encoded as `HtmlLinkInfo.Kind.MEMBER_TYPE_PARAMS`), and it was certainly never meant to be displayed with preformatted whitespace, so IMO the most likely reason to do this was to add a line break opportunity before we used `<wbr>`. Therefore, regardless of whether we decide to add a space or not, that flag and special case can be safely removed.

**Comment by RogerRiggs** [comment]
My preference is to keep the compact declarations compact (so no spaces between single character types).
The longer type parameters are more readable with spaces (after ,).

**Comment by nizarbenalla** [comment]
I'd like to be consistent with source formatting recommendations, and to have a uniform way of displaying output.
But I'm curious how many instances of single letters vs longer type parameters there are in the source code.

**Comment by hns** [comment]
I have to correct my previous comment, the newline characters in type parameters of generic methods are actually displayed as preformatted text when type parameters are very long, such as [in this method][1].

[1]: https://docs.oracle.com/en/java/javase/23/docs/api/java.net.http/java/net/http/HttpResponse.BodySubscribers.html#fromSubscriber(java.util.concurrent.Flow.Subscriber,java.util.function.Function)

**Comment by nizarbenalla** [comment]
For now, I will try to have no whitespace when we only have one letter parameters i.e `<K,V,T>` and add whitespace when we have type expressions as in `IntStream flatMapToInt(Function<? super T, ? extends IntStream> mapper)`

In case we have both, I'll still add whitespace.

**Comment by nizarbenalla** [comment]
I have dialed down this change a bit. Now only affecting uses of parameterized types, in which the type arguments are longer type expressions.

as in (new):
 Screenshot 2024-11-15 at 18 28 07
 
 vs (old)
 
 
Screenshot 2024-11-15 at 18 29 46


Type parameters in class headings will remain with no spaces.

I have updated the PR's body with a new link to the generated documentation.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlLinkFactory.java [hunk 1]
```diff
@@ -399,10 +400,16 @@ protected Content getTypeParameterLinks(HtmlLinkInfo linkInfo) {
             }
             links.add("<");
             boolean many = false;
+            boolean hasUsesOfParameterizedTypes =
```
Comment (pavelrappo) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlLinkFactory.java:403]: I think we should name this variable better. Other than that, looks good.

**Review by pavelrappo — CHANGES_REQUESTED** [review]

**Comment by hns** [comment]
To illustrate my previous comment, here are two screenshots of the `Spliterator.OfPrimitive` page heading:

1) in its current form without space

tp-no-space

2) with space added if bounds are present. 

tp-bounded-space

**Comment by nizarbenalla** [comment]
For visibility, here is the [generated Javadoc](https://cr.openjdk.org/~nbenalla/GeneratedDocs/8342808/docs/api/index.html) after this patch.

**Review by hns — CHANGES_REQUESTED** [review]
This looks good to me. The threshold of 8 characters is a bit arbitrary, but it seems reasonable. 

A few minor issues:
 - Since we don't really check for bounded type parameters but only for length I suggest renaming the boolean variable to something like `longTypeParams`.
 - The length threshold should be defined as a constant in a way that both `ClassWriter` and `HtmlLinkFactory` can use it.
 - The old `TYPE_PARAMS_MAX_INLINE_LENGTH` constant in `Signatures` isn't used anymore and should be removed.

**Comment by nizarbenalla** [comment]
I wasn't happy with using a magic number but was unsure on where to define it.
Looking at the source code, most constants are defined in Writers.

**Pull request closed without merging**